Use negative delay from first stop in a GTFS RT update in previous stop times when required

optionsome opened the pull request:
### Summary
If a GTFS RT trip update doesn't contain full stop sequence and the delay from the first stop is negative, this pr changes so that the same delay is set for previous stop times when required to ensure that the times are increasing.

Adds an optional `backwardsDelayPropagationType` "enum" parameter for the polling stop time updater. It has three options:
* "REQUIRED_NO_DATA" (default), only propagates delays backwards when it is required to ensure that the times are increasing and it sets the NO_DATA flag on the stops so these automatically updated times are not exposed through APIs and the stops at the beginning that have not received any updates, will not be shown as realtime updated 
* "REQUIRED", same as REQUIRED_NO_DATA, but NO_DATA flag is not automatically set on stops
* "ALWAYS", propagates delays backwards on stops with no estimates regardless if it's required or not, and NO_DATA flag is not set

### Issue
closes #2761

### Unit tests
Added tests.

### Code style
Have you followed the [suggested code style](https://github.com/opentripplanner/OpenTripPlanner/blob/dev-2.x/docs/Codestyle.md)?
Yes

### Documentation
Added documentation

### Changelog
From title

barbeau commented:
>Adds a propagateNegativeDelayBackwards parameter for the polling stop time updater but so this behavior can be disabled if it's important to keep the old behavior.

Could you reverse this, so it's disabled by default?

Technically propagating delays backwards in a trip violates the GTFS RT spec. 

[GTFS RT says](https://github.com/google/transit/blob/master/gtfs-realtime/spec/en/trip-updates.md#stop-time-updates):

>Example 2
>
>For the same trip instance, three [StopTimeUpdates](https://github.com/google/transit/blob/master/gtfs-realtime/spec/en/reference.md#StopTimeUpdate) are provided:
>
>* delay of 300 seconds for stop_sequence 3
>* delay of 60 seconds for stop_sequence 8
>* [ScheduleRelationship](https://github.com/google/transit/blob/master/gtfs-realtime/spec/en/reference.md/#enum-schedulerelationship) of NO_DATA for stop_sequence 10
>
>This will be interpreted as:
>
>* stop_sequences 1,2 have unknown delay.
>* stop_sequences 3,4,5,6,7 have delay of 300 seconds.
>* stop_sequences 8,9 have delay of 60 seconds.
>* stop_sequences 10,..,20 have unknown delay.

If I understand the behavior you're proposing here then stop_sequences 1 and 2 would have the delay of 300 seconds, which isn't what GTFS RT requires.

leonardehrenfried commented:
@barbeau Perhaps it's worth providing some context. The reason we are doing this is that for the routing algorithm to work it's a requirement that the stop times are always increasing. Many RT feeds don't have the data for the entire trip leading to non-increasing times in the OTP model and therefore the updates are rejected.

barbeau commented:
@leonardehrenfried Thanks for the context, that makes sense in terms of routing. So would these backwards propagated delays be visible to the end user (traveler)? The GTFS RT language above is trying to preserve the intent of the producer as seen by the end-user (traveler).

leonardehrenfried commented:
Sadly, they would be visible. Well, they would if we didn't exclude the entire trip.

optionsome commented:
We discussed this pr today in the developer meeting and came to the conclusion that this behavior should be put behind an enum configuration parameter so the estimates for the missing first stops can be either updater never, when required to have consistent times or always. We still thought the default behavior should be when required. It differs from the current implementation in a sense that the negative delays should only be backpropagated when the trip update would be ignored otherwise.

barbeau commented:
Would it be possible to have a flag on stops that contain backpropagated delays that would be available in the API? That would then allow UI clients to make the decision whether to show those backpropagated delays to the end user (traveler), which could preserve compliance with the GTFS RT spec.

optionsome commented:
We discussed that option also today but it didn't get too much support but it is possible. Do you have some use cases where you would need it?

barbeau commented:
Yes, I would use this flag in the [OneBusAway Android](https://github.com/OneBusAway/onebusaway-android) client to show the scheduled arrival/departure times for those stops to keep compliance with GTFS RT spec. I recall seeing instances in the past of OTP discarding TripUpdates in cities where OBA is used, which I believe was due to this issue.

Generally I also think it's a best practice to keep compliance with the spec (or at least allow clients to keep compliance).

leonardehrenfried commented:
Well, if you don't want this behaviour you could disable it altogether and keep the compliance, couldn't you? Or am I misunderstanding?

optionsome commented:
I feel like it would be confusing to display times to users times that are not increasing but I do get your point about following the specification.

hannesj commented:
The issue comes from the fact that OTP has stricter internal requirements, than what GTFS-RT guarantees, and we try being pragmatic about it by sometimes tweaking the data a bit, in order to get it to fit our model. 

Another option would be to keep the historical passed times from the previous updates, if they are dropped by the producer.

barbeau commented:
>Another option would be to keep the historical passed times from the previous updates, if they are dropped by the producer.

GTFS RT is stateless, so this approach wouldn't be compliant with the spec either.

>I feel like it would be confusing to display times to users times that are not increasing but I do get your point about following the specification.

To me this isn't as much about displaying the times as it is displaying real-time info vs. schedule info. If a producer chooses not to share a real-time update at a stop early in the trip there could be very valid reasons for this. Displaying a backpropagated delay on these stops vs. displaying the schedule convey two very different meanings to the end user (the first being "this is when the vehicle is predicted to arrive" and the second being "we don't know when the vehicle is arriving"). As a client I'd prefer to keep the original producer intent and show the schedule to the end user. But the only way I could do this is if there is a flag indicating that the stop contains backpropagated info.

In OneBusAway, this issue is also compounded by the fact that the [OBA server](https://github.com/OneBusAway/onebusaway-application-modules) complies with the GTFS RT spec on this topic and the OBA clients get their arrival time list from the OBA server and trip plan data from an OTP server. Therefore when viewing arrivals at a stop the rider would see schedule information for the early stops in the trip. If OTP doesn't offer clients the option to do the same, end users are going to see two different sets of information - the backpropagated delays in trip plans and schedule info when viewing lists of arrival times.

leonardehrenfried commented:
But if you want to keep the current behaviour, you can do that. @optionsome is proposing a feature that you can turn off. If you do that then everything stays at it is right now and the trip update is discarded. 

It sounds however that you would like the trip update to be applied but without backpropagation. This would be a new feature, wouldn't it?

What do you think about discussing this in person? Every Thursday we have a developer meeting in an American-friendly time zone: https://github.com/opentripplanner/OpenTripPlanner/blob/dev-2.x/CONTRIBUTING.md#primary-channels-of-communication

Would be great to have you with your detailed knowledge on board. We could also explain our practical problems as developers building travel planners.

hannesj commented:
> > 
> 
> To me this isn't as much about displaying the times as it is displaying real-time info vs. schedule info (with labels specifying it as such). If a producer chooses not to share a real-time update at a stop early in the trip there could be very valid reasons for this. Displaying a backpropagated delay on these stops vs. displaying the schedule convey two very different meanings to the end user (the first being "this is when the vehicle is predicted to arrive" and the second being "we don't know when the vehicle is arriving"). As a client I'd prefer to keep the original producer intent and show the schedule to the end user. But the only way I could do this is if there is a flag indicating that the stop contains backpropagated real-time info.

I think it is possible to set `ScheduleRelationship` the `NO_DATA`, and thus making the trip show the scheduled time at that stop. This is being worked on in https://github.com/opentripplanner/OpenTripPlanner/pull/3962

barbeau commented:
>It sounds however that you would like the trip update to be applied but without backpropagation. This would be a new feature, wouldn't it?

Conceptually, yes. But based on the constraints of the routing algorithm (times need to be increasing), it doesn't sound like this is possible. So I think the "new feature" beyond what's currently in this PR would be the flag on the stops indicating that they contained backpropagated times as discussed above.

>What do you think about discussing this in person? Every Thursday we have a developer meeting in an American-friendly time zone: https://github.com/opentripplanner/OpenTripPlanner/blob/dev-2.x/CONTRIBUTING.md#primary-channels-of-communication

Sure, that would be great. I can't attend tomorrow due to a conflict but I'll see if I can attend next week.

>I think it is possible to set ScheduleRelationship the NO_DATA, and thus making the trip show the scheduled time at that stop. This is being worked on in https://github.com/opentripplanner/OpenTripPlanner/pull/3962

Do you mean the producer should set this, or this is something OTP could be configured to do internally for the scenario of backpropagation?

optionsome commented:
> Do you mean the producer should set this, or this is something OTP could be configured to do internally for the scenario of backpropagation?
I think doing it internally in OTP could work but it should be configurable if these backpropagated times should or should not be shown instead of the scheduled times. Not sure if the same parameter as the one that controls the behaviour of the propagation should be used for that. Would it be important to expose through the API the difference between scheduled time due to no updates for any stops in the trip, NO_DATA in the update or that the times have been "fixed" by otp for routing purposes?

barbeau commented:
>I think doing it internally in OTP could work

Ok, that would be ideal then 👍 

>Not sure if the same parameter as the one that controls the behaviour of the propagation should be used for that.

Thinking about this more, from a general OTP user and client developer, IMHO the best case scenario is that the behavior out-of-the-box for OTP with minimal configuration should meet my expectations of the GTFS Realtime spec behavior requirements, which would be:
1. Consume valid GTFS Realtime TripUpdates (including those with negative delay values that start later in the trip)
2. The OTP API output should match the GTFS Realtime input and spec by default - so if no RT data was provided for stops early in the trip, then the API should reflect only schedule data available for these stops (and not the backpropagated delays).
3. Valid trip plans should be returned for trips that include negative delays mid-trip - this requires back-propagation internally for the routing algorithm to work but by default isn't exposed to the end user/API.

Then, if a OTP deployer wants to expose the backpropagated times via the API they should be able to do so via a configuration option.

>Would it be important to expose through the API the difference between scheduled time due to no updates for any stops in the trip, NO_DATA in the update or that the times have been "fixed" by otp for routing purposes?

If I'm understanding your question correctly, in the case we're discussing GTFS RT NO_DATA is equivalent to omitting a GTFS RT stop_time_update for a particular stop early in the trip. So from the OTP API perspective IMHO they can be represented the same way (no real-time available for this stop, schedule only information is available).

I believe the OTP plan response still looks something like this (?):

```json
        "legs": [
          {
            "startTime": 1618173471000,
            "endTime": 1618173900000,
            "departureDelay": 0,
            "arrivalDelay": 0,
            "realTime": false,
```

So I think the question is if we need separate variables there to represent the backpropagated delays?

hannesj commented:
> If I'm understanding your question correctly, in the case we're discussing GTFS RT NO_DATA is equivalent to omitting a GTFS RT stop_time_update for a particular stop early in the trip. So from the OTP API perspective IMHO they can be represented the same way (no real-time available for this stop, schedule only information is available).

Is there a section in the specification which describes how missing stops should be handled before the first update? I only see the following, which describes forward propagation. If not, I think we could raise an issue on the mailing list/Github for a clarification of the spec.

```
If one or more stops are missing along the trip the update is propagated to all subsequent stops.
This means that updating a stop time for a certain stop will change all subsequent stops in the 
absence of any other information.
```

optionsome commented:
> Is there a section in the specification which describes how missing stops should be handled before the first update? I only see the following, which describes forward propagation. If not, I think we could raise an issue on the mailing list/Github for a clarification of the spec.

I think this is the closes to it:
>You are allowed, but not required, to drop past stop times. Producers should not drop a past StopTimeUpdate if it refers to a stop with a scheduled arrival time in the future for the given trip (i.e. the vehicle has passed the stop ahead of schedule), as otherwise it will be concluded that there is no update for this stop.

barbeau commented:
>Is there a section in the specification which describes how missing stops should be handled before the first update?

Yes, they are effectively NO_DATA-  see my comment at https://github.com/opentripplanner/OpenTripPlanner/pull/4035#issuecomment-1080777796 for the quote from the spec.

optionsome commented:
We discussed this in the yesterday's meeting. We decided that the default behavior should be that the backwards propagation should be only done when it's required to keep the times increasing and we should mark those with the NO_DATA flag from https://github.com/opentripplanner/OpenTripPlanner/pull/3962 so that the stops can be indicated to not have been updated through the APIs. Optionally, it should also be possible to not flag the stops with NO_DATA and maybe to propagate the times regardless if it's required or not.

optionsome commented:
I implemented support for propagating departure delay to arrival of the first updated stop but now that I re-read the specification, I think both the arrival and departure times should be dropped simultaneously for the "past" stops. So I think I should drop that feature with the default configuration at least but should I include it for any other configuration?

optionsome commented:
This pr is now rewritten and rebased on https://github.com/opentripplanner/OpenTripPlanner/pull/3962 as the NO_DATA flags are now in use. I've updated the pr's summary. I wrote some tests for this but I haven't done too much manual testing yet. I might have to change how the times are exposed through the APIs when NO_DATA flag is set but I have to test how it works now. This pr is still draft waiting for the other pr to be merged.

optionsome commented:
I have updated the APIs now to not show delay if stop is canceled or flagged with NO_DATA and to show realtime=false if both the first and last stops have NO_DATA flags. If I have missed some API code that should also be updated, please let me know.

leonardehrenfried commented on src/test/java/org/opentripplanner/model/TimetableSnapshotTest.java:
```diff
@@ -105,12 +113,21 @@ public void testUpdate() {
 
         tripDescriptorBuilder.setTripId("1.1");
-        tripDescriptorBuilder.setScheduleRelationship(ScheduleRelationship.CANCELED);
+        tripDescriptorBuilder.setScheduleRelationship(ScheduleRelationship.SCHEDULED);
```
Could you perhaps explain why this is necessary? I think I'm missing something here. Should this not be a separate test case?

optionsome replied:
This tests through the `Timetable` class' `createUpdatedTripTimes` method. We don't use it for handling CANCELED updates as it's done through `TimetableSnapshotSource` class directly. Therefore, I removed the support for handling CANCELED trips in `createUpdatedTripTimes` and updated the tests that used it.

leonardehrenfried replied:
Ok, thanks for the explanation.

leonardehrenfried commented:
@optionsome You need to resolve the merge conflicts in this PR.
